src/terraform/install.sh

```diff
@@ -177,19 +218,94 @@ check_packages() {
     fi
 }
 
+# Function to fetch the version released prior to the latest version
+get_previous_version() {
+    local url=$1
+    local repo_url=$2
+    local variable_name=$3
+    prev_version=${!variable_name}
+    
+    output=$(curl -s "$repo_url");
+
+    # install jq if not exists
+    if ! type jq > /dev/null 2>&1; then
+        check_packages jq
+    fi
+    
+    message=$(echo "$output" | jq -r '.message')
+    
+    if [[ $message == "API rate limit exceeded"* ]]; then
+        echo -e "\nAn attempt to find latest version using GitHub Api Failed... \nReason: ${message}"
+        echo -e "\nAttempting to find latest version using GitHub tags."
+        find_prev_version_from_git_tags prev_version "$url" "tags/v"
+        declare -g ${variable_name}="${prev_version}"
+    else
+        echo -e "\nAttempting to find latest version using GitHub Api."
+        version=$(echo "$output" | jq -r '.tag_name')
+        declare -g ${variable_name}="${version#v}"
+    fi  
+    echo "${variable_name}=${!variable_name}"
+}
+
+get_github_api_repo_url() {
+    local url=$1
+    echo "${url/https:\/\/github.com/https:\/\/api.github.com\/repos}/releases/latest"
+}
+
+get_pkg_name() {
+    local input_string="$1"
+    local lowercase_input="${input_string,,}"  # Convert to lowercase
+    local suffix="_version"
+    local substring="${lowercase_input%$suffix*}"  # Remove suffix and everything after it
+    echo "$substring"
+}
+
+install_previous_version() {
+    given_version=$1
+    requested_version=${!given_version}
+    local URL=$2
+    local REPO_URL=$(get_github_api_repo_url "$URL")
+    local PKG_NAME=$(get_pkg_name "${given_version}")
+    echo -e "\n(!) Failed to fetch the latest artifacts for ${PKG_NAME} v${requested_version}..."
+    get_previous_version "$URL" "$REPO_URL" requested_version
+    echo -e "\nAttempting to install ${requested_version}"
+    declare -g ${given_version}="${requested_version#v}"
+    INSTALLER_FN="install_${PKG_NAME}"
```

Comment:
Can we simply pass the function name as arg to `install_previous_version`? 

I understand you are computing PKG_NAME by utilizing something like `TERRAFORM_VERSION` and fetching `terraform`. However, simply passing function name seems more cleaner to me.

Reply:
made changes. thanks..